### PR TITLE
changes to the cachekey to use tokenNotificationArgs 

### PR DESCRIPTION
**Pull request by jennyf19**
Next release of MSAL .NET will surface the cache key to use in the tokenNotificationArgs, so we'll use that instead. 

Making this draft PR for ease of testing. cc: @jmprieur 

#235

### CODE REVIEW

**Comment by jennyf19** [comment]
@pmaytak you could probably review this now, i don't expect much to change. If you want to build it and test it, although it's probably easier to wait until MSAL .NET releases.

File: src/Microsoft.Identity.Web/TokenAcquisition.cs [hunk 1]
```diff
@@ -283,29 +283,34 @@ public async Task<string> GetAccessTokenForAppAsync(IEnumerable<string> scopes)
...
                 {
-                    await app.RemoveAsync(account).ConfigureAwait(false);
-                    _tokenCacheProvider?.ClearAsync().ConfigureAwait(false);
+                    string? identifier = context.HttpContext.User.GetMsalAccountId();
+                    IAccount account = await app.GetAccountAsync(identifier).ConfigureAwait(false);
+
+                    if (account != null)
+                    {
+                        await app.RemoveAsync(account).ConfigureAwait(false);
+                        _tokenCacheProvider?.ClearAsync(userId).ConfigureAwait(false);
```
Comment (pmaytak) [inline src/Microsoft.Identity.Web/TokenAcquisition.cs:312]:
```suggestion
                        await _tokenCacheProvider?.ClearAsync(userId).ConfigureAwait(false);
```

File: src/Microsoft.Identity.Web/TokenAcquisition.cs [hunk 2]
```diff
@@ -283,29 +283,34 @@ public async Task<string> GetAccessTokenForAppAsync(IEnumerable<string> scopes)
...
-            {
-                string? identifier = context.HttpContext.User.GetMsalAccountId();
-                IAccount account = await app.GetAccountAsync(identifier).ConfigureAwait(false);
+                    foreach (var b2cAccount in b2cAccounts)
+                    {
+                        await app.RemoveAsync(b2cAccount).ConfigureAwait(false);
+                    }
 
-                if (account != null)
+                    _tokenCacheProvider?.ClearAsync(userId).ConfigureAwait(false);
```
Comment (pmaytak) [inline src/Microsoft.Identity.Web/TokenAcquisition.cs:302]:
```suggestion
                    await _tokenCacheProvider?.ClearAsync(userId).ConfigureAwait(false);
```

File: src/Microsoft.Identity.Web/TokenCacheProviders/IMsalTokenCacheProvider.cs [hunk 3]
```diff
@@ -21,7 +21,7 @@ public interface IMsalTokenCacheProvider
         /// <summary>
         /// Clear the cache.
         /// </summary>
-        /// <returns>A <see cref="Task"/> that represents a completed clear operation.</returns>
```
Comment (pmaytak) [inline src/Microsoft.Identity.Web/TokenCacheProviders/IMsalTokenCacheProvider.cs:24]: Keep this line.

File: src/Microsoft.Identity.Web/TokenCacheProviders/MsalAbstractTokenCacheProvider.cs [hunk 4]
```diff
@@ -121,16 +67,11 @@ protected virtual Task OnBeforeWriteAsync(TokenCacheNotificationArgs args)
         /// <summary>
         /// Clear the cache.
         /// </summary>
-        /// <returns>A <see cref="Task"/> that represents a completed clear operation.</returns>
```
Comment (pmaytak) [inline src/Microsoft.Identity.Web/TokenCacheProviders/MsalAbstractTokenCacheProvider.cs:124]: Keep this line.

File: src/Microsoft.Identity.Web/TokenCacheProviders/MsalAbstractTokenCacheProvider.cs [hunk 5]
```diff
@@ -121,16 +67,11 @@ protected virtual Task OnBeforeWriteAsync(TokenCacheNotificationArgs args)
...
-        public async Task ClearAsync()
+        /// <param name="homeAccountId">HomeAccountId for a user account in the cache.</param>
+        public async Task ClearAsync(string homeAccountId)
```
Comment (pmaytak) [inline src/Microsoft.Identity.Web/TokenCacheProviders/MsalAbstractTokenCacheProvider.cs:71]: Perhaps we can rename this method or specify in the summary that this clears user token cache? To make the purpose of this method clearer.

File: src/Microsoft.Identity.Web/TokenCacheProviders/Distributed/MsalDistributedTokenCacheAdapter.cs [hunk 6]
```diff
@@ -28,24 +27,19 @@ public class MsalDistributedTokenCacheAdapter : MsalAbstractTokenCacheProvider
...
-        /// <param name="httpContextAccessor"></param>
         /// <param name="memoryCache"></param>
         /// <param name="cacheOptions"></param>
         public MsalDistributedTokenCacheAdapter(
-                                            IOptions<MicrosoftIdentityOptions> microsoftIdentityOptions,
-                                            IHttpContextAccessor httpContextAccessor,
                                             IDistributedCache memoryCache,
-                                            IOptions<MsalDistributedTokenCacheAdapterOptions> cacheOptions)
-            : base(microsoftIdentityOptions, httpContextAccessor)
+                                            IOptions<DistributedCacheEntryOptions> cacheOptions)
```
Comment (pmaytak) [inline src/Microsoft.Identity.Web/TokenCacheProviders/Distributed/MsalDistributedTokenCacheAdapter.cs:34]: Why is this not `IOptions<MsalDistributedTokenCacheAdapterOptions> cacheOptions`?

Comment (jmprieur) [reply src/Microsoft.Identity.Web/TokenCacheProviders/Distributed/MsalDistributedTokenCacheAdapter.cs:34]: Good catch!

File: src/Microsoft.Identity.Web/TokenCacheProviders/Distributed/MsalDistributedTokenCacheAdapter.cs [hunk 7]
```diff
@@ -28,24 +27,19 @@ public class MsalDistributedTokenCacheAdapter : MsalAbstractTokenCacheProvider
...
-            : base(microsoftIdentityOptions, httpContextAccessor)
         {
             if (cacheOptions == null)
             {
                 throw new ArgumentNullException(nameof(cacheOptions));
             }
 
             _distributedCache = memoryCache;
-            _cacheOptions = cacheOptions.Value;
+            _cacheOptions = (MsalDistributedTokenCacheAdapterOptions)cacheOptions.Value;
```
Comment (pmaytak) [inline src/Microsoft.Identity.Web/TokenCacheProviders/Distributed/MsalDistributedTokenCacheAdapter.cs:42]:
```suggestion
            _cacheOptions = cacheOptions.Value;
```

**Pull request merged**